apps/emqx_mt/src/emqx_mt_limiter.erl

```diff
@@ -0,0 +1,185 @@
+    emqx_limiter_client_container:t().
+create_channel_client_container(Zone, ListenerId, Ns) ->
+    create_client_container(Zone, ListenerId, Ns, limiter_names()).
+
+create_tenant_limiter_group(Ns, Config) ->
+    LimiterConfigs = to_limiter_options(Config),
+    emqx_limiter:create_group(emqx_limiter_shared, tenant_group(Ns), LimiterConfigs).
+
+update_tenant_limiter_group(Ns, Config) ->
+    LimiterConfigs = to_limiter_options(Config),
+    emqx_limiter:update_group(tenant_group(Ns), LimiterConfigs).
+
+delete_tenant_limiter_group(Ns) ->
+    emqx_limiter:delete_group(tenant_group(Ns)).
+
+%%
+
+create_client_limiter_group(Ns, Config) ->
+    LimiterConfigs = to_limiter_options(Config),
+    emqx_limiter:create_group(emqx_limiter_exclusive, client_group(Ns), LimiterConfigs).
+
+update_client_limiter_group(Ns, Config) ->
+    LimiterConfigs = to_limiter_options(Config),
+    emqx_limiter:update_group(client_group(Ns), LimiterConfigs).
+
+delete_client_limiter_group(Ns) ->
+    emqx_limiter:delete_group(client_group(Ns)).
+
+%%------------------------------------------------------------------------------
+%% `emqx_channel:set_limiter_adjustment_fn' hook
+%%------------------------------------------------------------------------------
+
+adjust_limiter(#{tns := undefined}) ->
+    ignore;
+adjust_limiter(Context) ->
+    #{
+        zone := Zone,
+        listener_id := ListenerId,
+        tns := Ns
+    } = Context,
+    {ok, create_channel_client_container(Zone, ListenerId, Ns)}.
+
+%%------------------------------------------------------------------------------
+%% Internal fns
+%%------------------------------------------------------------------------------
+
+zone_group(Zone) ->
+    {zone, Zone}.
+
+listener_group(ListenerId) ->
+    {listener, ListenerId}.
+
+tenant_group(Ns) ->
+    {mt_tenant, Ns}.
+
+client_group(Ns) ->
+    {mt_client, Ns}.
+
+limiter_names() ->
+    [?MESSAGES_LIM_NAME, ?BYTES_LIM_NAME].
+
+to_limiter_options(Config) ->
+    lists:map(
+        fun(Name) ->
+            #{rate := Rate, burst := Burst} = maps:get(Name, Config),
+            {Name, emqx_limiter:config_from_rate_and_burst(Rate, Burst)}
+        end,
+        limiter_names()
+    ).
+
+create_client_container(Zone, ListenerId, Ns, Names) ->
+    Clients = lists:map(
+        fun(Name) ->
+            LimiterClient = create_limiter(Zone, ListenerId, Ns, Name),
+            {Name, LimiterClient}
+        end,
+        Names
+    ),
+    emqx_limiter_client_container:new(Clients).
+
+create_limiter(Zone, ListenerId, Ns, Name) ->
+    TenantLimiters = create_tenant_limiters(Zone, Ns, Name),
+    ClientLimiters = create_client_limiters(ListenerId, Ns, Name),
+    Clients = TenantLimiters ++ ClientLimiters,
+    emqx_limiter_composite:new(Clients).
+
+create_tenant_limiters(Zone, Ns, Name) ->
+    ZoneLimiterId = {zone_group(Zone), Name},
+    ZoneLimiterClient = emqx_limiter:connect(ZoneLimiterId),
+    case emqx_mt_config:get_tenant_limiter_config(Ns) of
```

Comment:
Curious

If the explicit ns configs appear later, do we agree to allow existing clients to remain unlimited by the new ns limits?

Reply:
Yes, I discussed this with @zmstone and reached this conclusion:

> we can keep it simple: only new ones will use the new config.
worst case, one can query all clients using the ns and kick them to force reconnect.